### PR TITLE
Add emp grenade

### PR DESCRIPTION
**Pull request by Slava0135**
## About the PR

Adds emp grenade. Right now it breaks lights and drains batteries around on trigger. 
In future it will disable some electronics, bots and radio, mess with chameleon clothes and etc.

**Media**

Peek 2023-03-04 13-34

**Changelog**

:cl:
- add: Added Emp Grenade to uplink (4 TC). Drains power from nearby batteries (laser gun, stun baton and etc.), turns off APCs and breaks lights around.

### CODE REVIEW

File: Content.Server/Emp/EmpSystem.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        foreach (var uid in _lookup.GetEntitiesInRange(coordinates.ToMap(EntityManager), range))
+        {
+            if (TryComp<BatteryComponent>(uid, out var battery))
+                battery.UseCharge(energyConsumption);
+            if (TryComp<PoweredLightComponent>(uid, out var light))
+                _poweredLight.TryDestroyBulb(light.Owner, light);
+        }
+        Spawn(EmpPulseEffectPrototype, coordinates);
+    }
```
Comment (metalgearsloth) [inline Content.Server/Emp/EmpSystem.cs:32]: This should be an event.

File: Content.Server/Emp/EmpSystem.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    private void HandleEmpTrigger(EntityUid uid, EmpOnTriggerComponent comp, TriggerEvent args)
+    {
+        EmpPulse(Transform(uid).Coordinates, comp.Range, comp.EnergyConsumption);
+        args.Handled = true;
```
Comment (metalgearsloth) [inline Content.Server/Emp/EmpSystem.cs:37]: Does this cause problems using entitycoordinates compare to attaching to grid / map, e.g. if the uid is deleted or in a container or whatever.

File: Content.Server/Emp/EmpSystem.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<EmpOnTriggerComponent, TriggerEvent>(HandleEmpTrigger);
+    }
+
+    public void EmpPulse(EntityCoordinates coordinates, float range, float energyConsumption)
+    {
+        foreach (var uid in _lookup.GetEntitiesInRange(coordinates.ToMap(EntityManager), range))
```
Comment (metalgearsloth) [inline Content.Server/Emp/EmpSystem.cs:24]: Doesn't this have an overload that takes entitycoordinates, also isn't ToMap obsolete.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Emp/EmpSystem.cs (hunk 3 above)
Comment (Slava0135) [reply Content.Server/Emp/EmpSystem.cs:24]: ToMap is not obsolete

**Comment by deltanedas** [comment]
when merged i will make ninja use it for emp ability, nice looking api :+1:

File: Content.Server/Emp/EmpSystem.cs (hunk 2 above)
Comment (Slava0135) [reply Content.Server/Emp/EmpSystem.cs:37]: yes it does, if you put it into backpack effect won't show up, or if you run with it in hand, effect will follow you

**Comment by Slava0135** [comment]
now affected entities spawn visual effect
Peek 2023-03-04 15-48

**Comment by Slava0135** [comment]
now also turns off APCs

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**